lgtm but tests are failing:
```
======================================================================
FAIL: test_next_subscription (corehq.apps.accounting.tests.test_models.TestSubscription)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/apps/accounting/tests/test_models.py", line 134, in test_next_subscription
    self.assertEqual(self.subscription.next_subscription, next_future_subscription)
AssertionError: <Subscription: Subscription to CommCare Advanced Edition (v1) for DOMAIN test. [Jan 10, 2018 - Jul 11, 2018]> != <Subscription: Subscription to CommCare Advanced Edition (v1) for DOMAIN test. [Jan 10, 2018 - Feb 11, 2018]>
----------------------------------------------------------------------
```

--- a/corehq/apps/accounting/models.py
+++ b/corehq/apps/accounting/models.py
@@ -6,6 +6,7 @@ import itertools
 
 import json_field
 from couchdbkit import ResourceNotFound
+from django.db.models import F
 from django.db.models.manager import Manager
 
 from corehq.util.quickcache import quickcache
@@ -1086,7 +1087,11 @@ class Subscription(models.Model):
         try:
             return Subscription.objects.filter(
                 subscriber=self.subscriber, date_start__gt=self.date_start
-            ).exclude(pk=self.pk).order_by('date_start')[0]
+            ).exclude(
+                pk=self.pk
+            ).exclude(
+                date_start=F('date_end')
+            ).order_by('date_start')[0]
         except (Subscription.DoesNotExist, IndexError):
             return None
 

--- a/corehq/apps/accounting/models.py
+++ b/corehq/apps/accounting/models.py
@@ -1,38 +1,31 @@
-from StringIO import StringIO
 import datetime
-from tempfile import NamedTemporaryFile
 from decimal import Decimal
 import itertools
+from StringIO import StringIO
+from tempfile import NamedTemporaryFile
 
-import json_field
-from couchdbkit import ResourceNotFound
-from django.db.models import F
-from django.db.models.manager import Manager
-
-from corehq.util.quickcache import quickcache
-from corehq.util.global_request import get_request
-from dimagi.ext.couchdbkit import DateTimeProperty, StringProperty, SafeSaveDocument, BooleanProperty
-import stripe
 
 from django.conf import settings
 from django.core.exceptions import ObjectDoesNotExist, ValidationError
 from django.core.validators import MaxValueValidator, MinValueValidator
 from django.db import models, transaction
+from django.db.models import F
+from django.db.models.manager import Manager
 from django.template.loader import render_to_string
-from django.utils.translation import ugettext_lazy as _
-from corehq.const import USER_DATE_FORMAT
-from corehq.util.view_utils import absolute_reverse
-from dimagi.utils.web import get_site_domain
 from django.utils.html import strip_tags
-
+from django.utils.translation import ugettext_lazy as _
 from django_prbac.models import Role
 
+import json_field
+import stripe
+
+from couchdbkit import ResourceNotFound
+from dimagi.ext.couchdbkit import DateTimeProperty, StringProperty, SafeSaveDocument, BooleanProperty
 from dimagi.utils.decorators.memoized import memoized
 from dimagi.utils.django.cached_object import CachedObject
+from dimagi.utils.web import get_site_domain
 
-from corehq.apps.hqwebapp.tasks import send_html_email_async
-from corehq.apps.users.models import WebUser
-
+from corehq.apps.accounting.emails import send_subscription_change_alert
 from corehq.apps.accounting.exceptions import (
     CreditLineError, AccountingError, SubscriptionAdjustmentError,
     SubscriptionChangeError, NewSubscriptionError, InvoiceEmailThrottledError,
@@ -40,6 +33,9 @@ from corehq.apps.accounting.exceptions import (
 )
 from corehq.apps.accounting.invoice_pdf import InvoiceTemplate
 from corehq.apps.accounting.signals import subscription_upgrade_or_downgrade
+from corehq.apps.accounting.subscription_changes import (
+    DomainDowngradeActionHandler, DomainUpgradeActionHandler,
+)
 from corehq.apps.accounting.utils import (
     get_privileges, get_first_last_days,
     get_address_from_invoice, get_dimagi_from_email_by_product,
@@ -50,11 +46,12 @@ from corehq.apps.accounting.utils import (
     log_accounting_info,
     quantize_accounting_decimal,
 )
-from corehq.apps.accounting.subscription_changes import (
-    DomainDowngradeActionHandler, DomainUpgradeActionHandler,
-)
-from corehq.apps.accounting.emails import send_subscription_change_alert
 from corehq.apps.domain.models import Domain
+from corehq.apps.hqwebapp.tasks import send_html_email_async
+from corehq.apps.users.models import WebUser
+from corehq.const import USER_DATE_FORMAT
+from corehq.util.quickcache import quickcache
+from corehq.util.view_utils import absolute_reverse
 
 integer_field_validators = [MaxValueValidator(2147483647), MinValueValidator(-2147483648)]
 

--- a/corehq/apps/accounting/tests/test_models.py
+++ b/corehq/apps/accounting/tests/test_models.py
@@ -121,6 +121,18 @@ class TestSubscription(BaseAccountingTest):
         self.subscription.save()
         self.assertEqual(1, len(Subscription.objects.filter(id=self.subscription.id)))
 
+    def test_next_subscription(self):
+        this_subscription_date_end = self.subscription.date_end
+        already_canceled_future_subscription, _ = generator.generate_domain_subscription_from_date(
+            this_subscription_date_end, self.account, self.domain.name,
+            subscription_length=0
+        )
+        next_future_subscription, _ = generator.generate_domain_subscription_from_date(
+            this_subscription_date_end, self.account, self.domain.name,
+            subscription_length=1
+        )
+        self.assertEqual(self.subscription.next_subscription, next_future_subscription)
+
     def tearDown(self):
         self.billing_contact.delete()
         self.dimagi_user.delete()

